### ISSUE TITLE
How to measure the BLEU of training/translation

### ISSUE THREAD

**Issue by xinyu-intel** [awslabs/sockeye#479]
Hi, 
I just trained a 8 layer rnn model and got the following result:
```
python -m sockeye.train -s corpus.tc.BPE.de \
                        -t corpus.tc.BPE.en \
                        -vs newstest2016.tc.BPE.de \
                        -vt newstest2016.tc.BPE.en \
                        --num-embed 512 \
                        --rnn-num-hidden 512 \
                        --rnn-attention-type dot \
                        --embed-dropout=0.2 \
                        --rnn-decoder-hidden-dropout=0.2 \
                        --max-seq-len 50 \
                        --decode-and-evaluate 500 \
                        --batch-size 128 \
			--batch-type sentence \
                        -o gnmt_model \
			--optimized-metric bleu \
                        --initial-learning-rate=0.0001 \
                        --learning-rate-reduce-num-not-improved=8 \
                        --learning-rate-reduce-factor=0.7 \
                        --weight-init xavier --weight-init-scale 3.0 \
                        --weight-init-xavier-factor-type avg \
			--lock-dir ~/.temp/ \
                        --num-layers 8:8 \
                        --device-ids 0 1
```
```
[2018-07-18:00:56:34:INFO:sockeye.training:fit] Training finished. Best checkpoint: 60. Best validation bleu: 0.746082
[2018-07-18:00:56:34:INFO:sockeye.utils:__exit__] Releasing GPU 1.
[2018-07-18:00:56:34:INFO:sockeye.utils:__exit__] Releasing GPU 0.
```
Evaluate translation:
```
[INFO:__main__] bleu	(s_opt)	chrf	(s_opt)
0.171	(-)	0.484	(-)
```
Is this bleu value correct? It is far away from the bleu value(20+) from the [paper](https://arxiv.org/abs/1712.05690).

By the way, how to use sockeye to build a GNMT model to align with the tensorflow's [config](https://github.com/tensorflow/nmt/blob/master/nmt/standard_hparams/wmt16_gnmt_8_layer.json)?

**Comment by tdomhan** [comment]
The exact training parameters from the models trained for the paper can be found here: 
https://github.com/awslabs/sockeye/tree/arxiv_1217/arxiv/code/rnn/sockeye 

How did you run the evaluation exactly? Ideally you will need to detokenize, merge BPE-based sub-words and then run `sacrebleu` on the output.

For GNMT you will need to additionally specify the following parameter: `--rnn-attention-in-upper-layers`. On the encoder side you will get the bi-directional RNN + a stack of uni-direction RNNs by default.

I hope that helps.

**Comment by xinyu-intel** [comment]
@tdomhan Thank you. I just use `python -m sockeye.evaluate -i -r` to get the score. Can you please tell me how to detokenize and merge?
By the way, what is the `--rnn-attention-type` for gnmt, dot or mlp? And should I set `--encoder rnn --decoder rnn`  together with `--rnn-attention-in-upper-layers` for it?

**Comment by tdomhan** [comment]
good point. Depends on which version of Sockeye your are using (master vs the arxiv_1217 branch). For the master branch you actually need to set `--encoder rnn --decoder rnn`.  `sockeye.evaluate` assumes that you tokenized the test data, whereas `sacrebleu` does not. For that reason you should ideally use the latter for WMT (and the former if you want to run a custom tokenizer for other datasets). Details for how to run this can be found here:
https://github.com/awslabs/sockeye/blob/arxiv_1217/arxiv/code/rnn/sockeye/test.sh

**Comment by tdomhan** [comment]
For the attention mechanism you should use `mlp` for a GNMT model. It is the default value though.

**Comment by xinyu-intel** [comment]
@tdomhan Okay, thank you for supporting. Can I add such config of GNMT in `https://github.com/awslabs/sockeye/blob/master/contrib/autopilot/models.py`?

**Comment by fhieber** [comment]
I think it would be useful to have a representative config for an RNN-based model somewhere in the repo. The autopilot model configurations sound like a good spot to have this.

**Comment by xinyu-intel** [comment]
@tdomhan thx but one more question, I've detokenized translate output to `de-en.output.detok` by using moses. How to use sacrebleu to evaluate it with the reference `newstest2016.tc.BPE.en`?

**Comment by xinyu-intel** [comment]
@tdomhan 
i've just use `cat de-en.output.detok | newstest2016.tc.en` and got the following result:
```
BLEU+case.mixed+numrefs.1+smooth.exp+tok.13a+version.1.2.7 = 3.99 17.1/6.0/2.4/1.0 (BP = 1.000 ratio = 2.781 hyp_len = 192128 ref_len = 69086)
```
Which number should be align with the score in paper?

**Comment by mjpost** [comment]
The purpose of sacrebleu is to avoid touching the reference all together. After postprocessing your output, you should just type:

    cat de-en.output.detok | sacrebleu -t wmt16 -l de-en

This will download the reference for you and give you a score.

**Comment by xinyu-intel** [comment]
@mjpost thx, does this `BLEU+case.mixed+numrefs.1+smooth.exp+tok.13a+version.1.2.7 = 3.99 17.1/6.0/2.4/1.0 (BP = 1.000 ratio = 2.781 hyp_len = 192128 ref_len = 69086)`mean bleu is 3.99?

**Comment by mjpost** [comment]
Yes. You can have it output just the score (minus all the details) with the `-b` flag.

Something is clearly wrong since this is a very low BLEU score indicating you have garbage output.

**Comment by xinyu-intel** [comment]
okay, but this is trained based on the GNMT config which @tdomhan recommended before:
```
python -m sockeye.train -s corpus.tc.BPE.de \
                        -t corpus.tc.BPE.en \
                        -vs newstest2016.tc.BPE.de \
                        -vt newstest2016.tc.BPE.en \
                        --num-embed 512 \
			--encoder rnn \
			--decoder rnn \
                        --rnn-num-hidden 512 \
			--rnn-attention-in-upper-layers \
                        --rnn-attention-type mlp \
                        --embed-dropout=0.2 \
                        --rnn-decoder-hidden-dropout=0.2 \
                        --max-seq-len 50 \
                        --decode-and-evaluate 500 \
                        --batch-size 128 \
			--batch-type sentence \
                        -o gnmt_model_rnn \
			--optimized-metric bleu \
                        --initial-learning-rate=0.0001 \
                        --learning-rate-reduce-num-not-improved=8 \
                        --learning-rate-reduce-factor=0.7 \
                        --weight-init xavier --weight-init-scale 3.0 \
                        --weight-init-xavier-factor-type avg \
			--lock-dir ~/.temp/ \
                        --num-layers 8:8 \
                        --device-ids 0 1
```
How can I modify this to reach a better bleu?

**Comment by tdomhan** [comment]
Which version of Sockeye are you using? The parameters you specified also don't all match the ones in the training script (https://github.com/awslabs/sockeye/blob/arxiv_1217/arxiv/code/rnn/sockeye/train.sh).

For example the maximum length of 50 is fairly low (use 100 instead). You probably want to optimize for perplexity (by removing `--optimized-metric`). You might want to add layer norm `--layer-normalization`.

Can you show the head of the decoding output file just for checking whether it's a model or a BLEU calculation issue?

**Comment by xinyu-intel** [comment]
hi, the sockeye version is 1.18.10 and the head of decoding outputfile is:
```
input=Obama empfängt Netanyahu  output=Obama is a Netanyahu     input_tokens=6  output_tokens=7 translation_time=0.0624
```
I tried to use two P100s ti train https://github.com/awslabs/sockeye/blob/arxiv_1217/arxiv/code/rnn/sockeye/train.sh
and it came across OOM.

**Comment by tdomhan** [comment]
The reason you see OOM is the change of default parameters from Sockeye `1.12.3` from the arxiv branch and your Sockeye version so that the default parameters don't match up. As you correctly pointed out above `train.sh` used with the newer Sockeye would actually train a Transformer model.

That said, I would still change the parameters mentioned above.

Is the output that you gave above what you pass into sacrebleu? How did you generate them (CLI args)?

The reason I'm asking is that they contain meta information (such as the number of input/output tokens). Sacrebleu however expects to only get the raw target side.

**Comment by xinyu-intel** [comment]
@tdomhan i use this:
```
cat $OUTPUT | \
    sed 's/\@\@ //g' | \
    $MOSES/scripts/tokenizer/detokenizer.perl -q -l en | \
    tee $OUTPUT.detok | \
    sacrebleu -b ./newstest2016.tc.en > $OUTPUT.bleu
```

**Comment by mjpost** [comment]
You shouldn't be providing the reference to sacrebleu. Your command should be this:

```bash
cat $OUTPUT | \
    sed 's/\@\@ //g' | \
    $MOSES/scripts/tokenizer/detokenizer.perl -q -l en | \
    tee $OUTPUT.detok | \
    sacrebleu -b -t wmt16 -l de-en > $OUTPUT.bleu
```

**Comment by xinyu-intel** [comment]
@mjpost thanks a lot!

**Comment by mjpost** [comment]
Did this fix your issue? Can you update with the results?

**Comment by xinyu-intel** [comment]
@mjpost yep, I got the correct bleu but i'm still working on training a good model.

**Comment by xinyu-intel** [comment]
@tdomhan yep, I removed the benchmark and got the correct answer. Thanks a lot!

**Comment by tdomhan** [comment]
oh ok, got it. What BLEU scores do you get now?

**Comment by xinyu-intel** [comment]
@tdomhan I have trained another model for near one week, the BLEU score has reached 27.23 now and the perplexity is around 8.0. Below are configs and training curve. When will this training progress stop since it seems almost converge.

```
python -m sockeye.train -s corpus.tc.BPE.de \
                        -t corpus.tc.BPE.en \
                        -vs newstest2016.tc.BPE.de \
                        -vt newstest2016.tc.BPE.en \
                        --num-embed 256 \
			--encoder rnn \
			--decoder rnn \
                        --rnn-num-hidden 512 \
			--rnn-attention-in-upper-layers \
                        --rnn-attention-type dot \
                        --embed-dropout=0.2 \
                        --rnn-decoder-hidden-dropout=0.2 \
                        --max-seq-len 100 \
                        --decode-and-evaluate 1000 \
                        --batch-size 128 \
			--batch-type sentence \
                        -o gnmt_model_rnn_dot_256_new \
                        --initial-learning-rate=0.0001 \
                        --learning-rate-reduce-num-not-improved=8 \
                        --learning-rate-reduce-factor=0.7 \
			--layer-normalization \
                        --weight-init xavier --weight-init-scale 3.0 \
                        --weight-init-xavier-factor-type avg \
			--lock-dir ~/.temp/ \
                        --num-layers 8:8 \
                        --device-ids 0 1
```
![image](https://user-images.githubusercontent.com/17897736/43385014-3d17aed0-9412-11e8-91e8-96303bdec056.png)

**Comment by fhieber** [comment]
closing as this seems resolved. Feel free to reopen if there are more related questions.
